dag_in_context/src/add_context.rs

```diff
@@ -3,25 +3,37 @@
 //! by remembering the most recent context (ex. DoWhile or If).
 //! Mantains the sharing invariant (see restore_sharing_invariant) by using a cache.
 
+use egglog::Term;
 use std::collections::HashMap;
 
 use crate::{
+    print_with_intermediate_helper,
     schema::{Assumption, Expr, RcExpr, TreeProgram},
     schema_helpers::AssumptionRef,
+    to_egglog::TreeToEgglog,
 };
 
-struct ContextCache {
+pub struct ContextCache {
     with_ctx: HashMap<(*const Expr, AssumptionRef), RcExpr>,
     symbol_gen: HashMap<(*const Expr, AssumptionRef), String>,
+    /// Information for generating the unions for loop contexts
+    placeholder: usize,
```

Comment:
edit after reading through the rest of the code:
Here's how I think this is working (but please correct me if I'm wrong!!)
When we add context to a loop, first we create a new context (I'm not sure why it's an `InFunc` assumption) and use that to add context to the pred and body.
Now we have two contexts: the Loop ctx, and the dummy loop context that we made. We add this pair to `loop_context_unions` so that we can add egglog code to union them later. This way, the loop with context doesn't use the original `pred_and_body` term at all (this was the shadow realm).
This is in contrast to the old way of doing it where the loop assumption would point to the `pred_and_body` without context nodes.

Ok, so hopefully that's at least mostly right. If so, I think some of the names could be improved. For example, it's not really clear what `placeholder` is for. I think it's like a counter kind of thing to make sure we have unique names for each of these loop ctx things. I think the name should allude to that (`idx`, `ctr`, something like that). Also, you sometimes re-use the name `placeholder` to refer to the ctx itself. I think it'd be better to not overload the term in that way.

Reply:
Yep, that's all correct. I'll change the name and add a comment to the place where I used `InFunc` (it's a hack to avoid changing the schema).

Reply:
I'm going to use placeholder everywhere to refer to the "dummy" loop context that we will union later. (The problem is that "dummy" contexts already mean something else.)